core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/queue/ElasticSearchDurationChecker.scala

```diff
@@ -0,0 +1,252 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.openwhisk.core.scheduler.queue
+
+import akka.actor.ActorSystem
+import com.sksamuel.elastic4s.http.ElasticDsl._
+import com.sksamuel.elastic4s.http.{ElasticClient, ElasticProperties, NoOpRequestConfigCallback}
+import com.sksamuel.elastic4s.{ElasticDate, ElasticDateMath, Seconds}
+import org.apache.openwhisk.common.Logging
+import org.apache.openwhisk.core.ConfigKeys
+import org.apache.openwhisk.core.entity.WhiskActionMetaData
+import org.apache.openwhisk.spi.Spi
+import pureconfig.loadConfigOrThrow
+import spray.json.{JsArray, JsNumber, JsValue, RootJsonFormat, deserializationError, _}
+
+import scala.concurrent.Future
+import scala.concurrent.duration.FiniteDuration
+import scala.language.implicitConversions
+import scala.util.{Failure, Try}
+
+trait DurationChecker {
+  def checkAverageDuration(invocationNamespace: String, actionMetaData: WhiskActionMetaData)(
+    callback: DurationCheckResult => DurationCheckResult): Future[DurationCheckResult]
+}
+
+case class DurationCheckResult(averageDuration: Option[Double], hitCount: Long, took: Long)
+
+object ElasticSearchDurationChecker {
+  val FilterAggregationName = "filterAggregation"
+  val AverageAggregationName = "averageAggregation"
+
+  implicit val serde = new ElasticSearchDurationCheckResultFormat()
+
+  def getFromDate(timeWindow: FiniteDuration): ElasticDateMath =
+    ElasticDate.now minus (timeWindow.toSeconds.toInt, Seconds)
+}
+
+class ElasticSearchDurationChecker(private val client: ElasticClient, val timeWindow: FiniteDuration)(
+  implicit val actorSystem: ActorSystem,
+  implicit val logging: Logging)
+    extends DurationChecker {
+  import ElasticSearchDurationChecker._
+  import org.apache.openwhisk.core.database.elasticsearch.ElasticSearchActivationStore.generateIndex
+
+  implicit val ec = actorSystem.getDispatcher
+
+  override def checkAverageDuration(invocationNamespace: String, actionMetaData: WhiskActionMetaData)(
+    callback: DurationCheckResult => DurationCheckResult): Future[DurationCheckResult] = {
+    val index = generateIndex(invocationNamespace)
+    val fqn = actionMetaData.fullyQualifiedName(false)
+    val fromDate = getFromDate(timeWindow)
+
+    logging.info(this, s"check average duration for $fqn in $index for last $timeWindow")
+
+    actionMetaData.binding match {
+      case Some(binding) =>
+        client
```

Comment:
the Some and None cases can call a helper function since the only difference in the query is the `List` to match on and pass that `List` as a param.

Reply:
updated accordingly